File: sdk/typescript/worker.ts
/** Rheovela TypeScript worker SDK (no runtime deps; Node 18+ fetch). */
export interface WorkItem {
  id: string;
  instance_id: string;
  activity_id: string;
  state: string;
}

export type WorkResult = { status: "success" | "failure"; error?: string };

export type WorkFn = (item: WorkItem) => Promise<WorkResult>;

export class WorkerError extends Error {
  status: number;
  constructor(status: number, message: string) {
    super(message);
    this.status = status;
    this.name = "WorkerError";
  }
}

export class Client {
  constructor(private baseUrl: string) {}

  private async req(method: string, path: string, body?: unknown): Promise<any> {
    const res = await fetch(this.baseUrl + path, {
      method,
      headers: { "Content-Type": "application/json" },
      body: body ? JSON.stringify(body) : undefined,
    });
    const text = await res.text();
    if (!res.ok) throw new WorkerError(res.status, text);
    return text ? JSON.parse(text) : {};
  }

  poll(instanceId?: string): Promise<WorkItem[]> {
    const path = instanceId
      ? `/api/v1/work-items?instance=${encodeURIComponent(instanceId)}`
      : "/api/v1/work-items";
    return this.req("GET", path);
  }

  claim(id: string, lease?: string): Promise<{ token: string; lease: string }> {
    return this.req("POST", `/api/v1/work-items/${id}/claim`, { lease });
  }

  heartbeat(id: string, token: string, lease?: string): Promise<{ ok: boolean }> {
    return this.req("POST", `/api/v1/work-items/${id}/heartbeat`, { token, lease });
  }

  complete(id: string, token: string): Promise<{ ok: boolean }> {
    return this.req("POST", `/api/v1/work-items/${id}/complete`, { token });
  }

  fail(id: string, token: string, error: string): Promise<{ ok: boolean }> {
    return this.req("POST", `/api/v1/work-items/${id}/fail`, { token, error });
  }
}

export class Worker {
  constructor(
    private client: Client,
    private fn: WorkFn,
    private lease = "30s",
  ) {}

  async processOnce(instanceId?: string): Promise<number> {
    let processed = 0;
    for (const item of await this.client.poll(instanceId)) {
      let token: string;
      try {
        token = (await this.client.claim(item.id, this.lease)).token;
      } catch {
        continue; // claimed by another worker
      }
      const res = await this.fn(item);
      try {
        if (res.status === "success") await this.client.complete(item.id, token);
        else await this.client.fail(item.id, token, res.error ?? "failure");
        processed++;
      } catch {
        /* stale token / expired lease */
      }
    }
    return processed;
  }
}

File: sdk/python/worker.py
"""Rheovela Python worker SDK (stdlib only).

Uses the core Worker HTTP API (rheo serve). Minimal, dependency-free.
"""
import json
import urllib.request
import urllib.error


class WorkerError(Exception):
    pass


class Client:
    def __init__(self, base_url):
        self.base_url = base_url.rstrip("/")

    def _request(self, method, path, body=None):
        url = self.base_url + path
        data = json.dumps(body).encode() if body is not None else None
        req = urllib.request.Request(url, data=data, method=method,
                                     headers={"Content-Type": "application/json"})
        try:
            with urllib.request.urlopen(req) as resp:
                return json.loads(resp.read().decode())
        except urllib.error.HTTPError as e:
            try:
                msg = e.read().decode()
            finally:
                e.close()
            raise WorkerError(f"{e.code}: {msg}") from e

    def poll(self, instance_id=None):
        path = "/api/v1/work-items"
        if instance_id:
            path += "?instance=" + instance_id
        return self._request("GET", path)

    def claim(self, work_item_id, lease="30s"):
        return self._request("POST", f"/api/v1/work-items/{work_item_id}/claim",
                             {"lease": lease})

    def heartbeat(self, work_item_id, token, lease="30s"):
        return self._request("POST", f"/api/v1/work-items/{work_item_id}/heartbeat",
                             {"token": token, "lease": lease})

    def complete(self, work_item_id, token):
        return self._request("POST", f"/api/v1/work-items/{work_item_id}/complete",
                             {"token": token})

    def fail(self, work_item_id, token, error):
        return self._request("POST", f"/api/v1/work-items/{work_item_id}/fail",
                             {"token": token, "error": error})


class Worker:
    def __init__(self, client, fn, lease="30s"):
        self.client = client
        self.fn = fn  # fn(item) -> ("success"|"failure", error_message_or_None)
        self.lease = lease

    def process_once(self, instance_id=None):
        processed = 0
        for item in self.client.poll(instance_id):
            try:
                claim = self.client.claim(item["id"], self.lease)
                token = claim["token"]
            except WorkerError:
                continue  # claimed by another worker
            status, err = self.fn(item)
            try:
                if status == "success":
                    self.client.complete(item["id"], token)
                else:
                    self.client.fail(item["id"], token, err or "failure")
                processed += 1
            except WorkerError:
                pass  # lease expired / stale token
        return processed
